libyara/rules.c

```diff
@@ -199,6 +199,54 @@ YR_API int yr_rules_scan_mem_blocks(
 }
 
 
+YR_API int yr_rules_scan_mem_init(
```

Comment:
Okay, I removed `yr_rules_scan_mem_init()` and `yr_rules_scan_mem_fini()`. My plan was to try and preserve and re-use their business logic as part of `yr_rules_scan_mem()` ... but you are right that the names could be more intention revealing. That's certainly something that I found confusing (that so many function names look so similar, that I found it gets in the way a bit regarding code comprehension, at least for me personally) when starting with libyara myself, so I can definitely relate :-)

Now that the instrumentation is in there then I'd love to refactor many of the names and make them more intention revealing, but that's a totally different kettle of fish, and I'm not sure if it's wise for the project. So I'm very happy that you are giving me this guidance with these comments.

I also stopped using `yr_rules_scan_mem()` in `test-async.c`. And in `test-rules.c` I originally added a 3rd loop over the tests, so that `yr_rules_scan_mem()` and the default iterator is still tested, but in addition to the other two loops which test aspects of the 'blockification' using the test iterator.

So `matches_blob()` in `util.c` now contains the original business logic with default iterator:

```
    // Call   yr_rules_scan_mem()                      <- Create & config scanner, and
    // calls  yr_scanner_scan_mem()                    <- Create default iterator, and
    // calls  yr_scanner_scan_mem_blocks()             <- Scan
```

As well as the 'blockification' business logic using the test iterator:

```
    // Call   yr_scanner_create()                      <- Create scanner
    // Call   yr_scanner_set_*()                       <- Config scanner
    // Call  _yr_test_single_or_multi_block_scan_mem() <- Create test iterator, and
    // calls  yr_scanner_scan_mem_blocks()             <- Scan
    // Call   yr_scanner_destroy()                     <- Destroy scanner
```

I agree it's looking better! :-)